### ISSUE TITLE
Support proto3 optional fields

### ISSUE THREAD

**Issue by prestonvanloon** [bazelbuild/rules_go#2673]
### What version of rules_go are you using?

Master 312a801084712314ea7da6f1c226306af58eb097

### What version of gazelle are you using?

N/A or latest

### What version of Bazel are you using?

3.6.0

### Does this issue reproduce with the latest releases of all the above?

yes

### What operating system and processor architecture are you using?

linux amd64

### Any other potentially useful information about your toolchain?

The "optional" label was removed when proto3 was released. However, it was introduced to proto3 in version 3.12.0.
See more info here: https://github.com/protocolbuffers/protobuf/blob/e2cc2de304f8bc33b91ab18fa0b7cef8aaf26c8b/docs/field_presence.md

### What did you do?

First, I updated `com_google_protobuf` to v3.13.0 (requires at least v3.12.0 for optional support in proto3).

So I am trying to use a proto3 message with an optional field. For example:

```protobuf
// tests/core/go_proto_library/optional.proto
syntax = "proto3";

option go_package = "github.com/bazelbuild/rules_go/tests/core/go_proto_library/protos";

message Optional {
  optional uint64 a = 1;
}
```

When I run this with protoc, then it works as expected when I provide the flag `--experimental_allow_proto3_optional`.

```
➜  rules_go git:(master) ✗ /home/preston/.local/bin/protoc --version   
libprotoc 3.13.0
➜  rules_go git:(master) ✗ /home/preston/.local/bin/protoc --experimental_allow_proto3_optional tests/core/go_proto_library/optional.proto --go_out=/tmp/ 

# produces expected file /tmp/github.com/bazelbuild/rules_go/tests/core/go_proto_library/protos/optional.pb.go
```

I added this proto and a basic test

```diff
--- a/tests/core/go_proto_library/BUILD.bazel
+++ b/tests/core/go_proto_library/BUILD.bazel
@@ -123,6 +123,25 @@ go_test(
     deps = [":protos_go_proto"],
 )
 
+go_test(
+    name = "optional_test",
+    srcs = ["optional_test.go"],
+    deps = [":optional_go_proto"],
+)
+
+go_proto_library(
+    name = "optional_go_proto",
+    importpath = "github.com/bazelbuild/rules_go/tests/core/go_proto_library/protos",
+    protos = [
+        ":optional_proto",
+    ]
+)
+
+proto_library(
+    name = "optional_proto",
+    srcs = ["optional.proto"],
+)
+
 go_proto_library(
     name = "protos_go_proto",
     importpath = "github.com/bazelbuild/rules_go/tests/core/go_proto_library/protos",
```

```go
// tests/core/go_proto_library/optional_test.go
package protos_test

import (
        "testing"

        "github.com/bazelbuild/rules_go/tests/core/go_proto_library/protos"
)

func use(interface{}) {}

func TestProtosWithOptional(t *testing.T) {
        use(protos.Optional{})
}
```

At this point, I am stuck / lost on where to add the flag `--experimental_allow_proto3_optional` to protoc. 
This issue might be a rules_proto issue, so I have also filed https://github.com/bazelbuild/rules_proto/issues/71. 

### What did you expect to see?

It works as intended.

### What did you see instead?

```
tests/core/go_proto_library/optional.proto: This file contains proto3 optional fields, but --experimental_allow_proto3_optional was not set.
```
Full log below
```
bazel test //tests/core/go_proto_library:optional_test
INFO: Invocation ID: c3339e91-610f-459a-92fe-97884f266bbc
INFO: Streaming build results to: https://builds.prylabs.net/invocation/c3339e91-610f-459a-92fe-97884f266bbc
DEBUG: Rule 'com_google_protobuf' indicated that a canonical reproducible form can be obtained by modifying arguments sha256 = "9b4ee22c250fe31b16f1a24d61467e40780a3fbb9b91c3b65be2a376ed913a1a"
DEBUG: Repository com_google_protobuf instantiated at:
  no stack (--record_rule_instantiation_callstack not enabled)
Repository rule http_archive defined at:
  /home/preston/.cache/bazel/_bazel_preston/86fca68c03a78f7c75ec8774faa99d1e/external/bazel_tools/tools/build_defs/repo/http.bzl:336:31: in <toplevel>
INFO: Analyzed target //tests/core/go_proto_library:optional_test (3 packages loaded, 29 targets configured).
INFO: Found 1 test target...
ERROR: /tmp/rules_go/tests/core/go_proto_library/BUILD.bazel:140:14: Generating Descriptor Set proto_library //tests/core/go_proto_library:optional_proto failed (Exit 1) protoc failed: error executing command bazel-out/host/bin/external/com_google_protobuf/protoc '--descriptor_set_out=bazel-out/k8-fastbuild/bin/tests/core/go_proto_library/optional_proto-descriptor-set.proto.bin' ... (remaining 5 argument(s) skipped)

Use --sandbox_debug to see verbose messages from the sandbox
tests/core/go_proto_library/optional.proto: This file contains proto3 optional fields, but --experimental_allow_proto3_optional was not set.
Target //tests/core/go_proto_library:optional_test failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 0.696s, Critical Path: 0.12s
INFO: 6 processes: 6 internal.
//tests/core/go_proto_library:optional_test                     FAILED TO BUILD

Executed 0 out of 1 test: 1 fails to build.
INFO: Streaming build results to: https://builds.prylabs.net/invocation/c3339e91-610f-459a-92fe-97884f266bbc
FAILED: Build did NOT complete successfully
```

**Comment by jayconrod** [comment]
protoc is invoked twice: once in `proto_library` to generate the binary descriptor consumed by other languages, and once in `go_proto_library` when using the Go plugin. It looks like `--experimental_allow_proto3_optional` would need to be passed to protoc in both commands, so this would need to be fixed in both rules_go and rules_proto.

Let's wait to see what the rules_proto folks say. If there's a standard way to do this, then rules_go should follow that.

**Comment by glerchundi** [comment]
I'm not quite sure but I think I've been bitten by this issue as well: https://github.com/grpc-ecosystem/grpc-gateway/pull/1834

**Comment by 0xKiwi** [comment]
Hello @jayconrod ! Been having issues integrating proto3 optional fields similar to @prestonvanloon . We [filed an issue with rules_proto](https://github.com/bazelbuild/rules_proto/issues/71), and they said the flag is [not currently exposed to Starlark](https://github.com/bazelbuild/rules_proto/issues/71#issuecomment-716434228). 

Any recommendations on how to move forward on this?

**Comment by jayconrod** [comment]
@0xKiwi Sorry, I don't think there's a straightforward solution unless those flags are exposed to Starlark. It sounds like you'd need to pass ` --protocopt=--experimental_allow_proto3_optional` on the command line, then also define a `go_proto_compiler` that invokes `protoc` with that flag.

**Comment by prestonvanloon** [comment]
Protoc 3.5.0 has flipped the flag for experimental allow proto3 optional to be on by default. We can now support these optional fields by using a recent version of protoc.

**Comment by powerbill** [comment]
> Protoc 3.5.0 has flipped the flag for experimental allow proto3 optional to be on by default. We can now support these optional fields by using a recent version of protoc.

I was confused until I went to your [PR](https://github.com/bazelbuild/rules_go/pull/2788): support for optional fields in protoc is available in 3.15 or later, correct?
